diff --git a/haskell/cabal.bzl b/haskell/cabal.bzl
index 0c17e93..2fb5b5b 100644
--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -118,8 +118,19 @@ def _prepare_cabal_inputs(hs, cc, dep_info, cc_info, package_id, tool_inputs, to
     args = hs.actions.args()
     package_databases = dep_info.package_databases
     extra_headers = cc_info.compilation_context.headers
-    extra_include_dirs = cc_info.compilation_context.includes
-    extra_lib_dirs = [file.dirname for file in ghci_extra_libs.to_list()]
+    extra_include_dirs = depset(transitive = [
+        cc_info.compilation_context.includes,
+        cc_info.compilation_context.quote_includes,
+        cc_info.compilation_context.system_includes,
+    ])
+    extra_lib_dirs = [
+        file.dirname
+        for file in ghci_extra_libs.to_list()
+        # Exclude Haskell libraries, as these are already covered by
+        # package-dbs. This is to avoid command-line length overflow on
+        # collect2.
+        if not file.basename.startswith("libHS")
+    ]
     args.add_all([package_id, setup, cabal.dirname, package_database.dirname])
     args.add("--flags=" + " ".join(flags))
     args.add("--")
